cmd/assetsvc/postgresql_utils.go

```diff
@@ -235,34 +213,77 @@ func (m *postgresAssetManager) getChartFilesWithFallback(namespace, filesID stri
+	whereClauses := []string{}
+	whereQueryParams := []interface{}{}
+	whereQuery := ""
+
+	if cq.namespace != dbutils.AllNamespaces {
+		whereQueryParams = append(whereQueryParams, cq.namespace, m.GetKubeappsNamespace())
+		whereClauses = append(whereClauses, fmt.Sprintf(
+			"(repo_namespace = $%d OR repo_namespace = $%d)", len(whereQueryParams)-1, len(whereQueryParams),
+		))
 	}
-	dbQuery := fmt.Sprintf("SELECT info FROM %s WHERE %s ORDER BY info ->> 'ID' ASC", dbutils.ChartTable, strings.Join(clauses, " AND "))
-	charts, err := m.QueryAllCharts(dbQuery, queryParams...)
-	if err != nil {
-		return nil, err
+	if cq.chartName != "" {
+		whereQueryParams = append(whereQueryParams, cq.chartName)
+		whereClauses = append(whereClauses, fmt.Sprintf(
+			"(info->>'name' = $%d)", len(whereQueryParams),
+		))
 	}
-	result := []*models.Chart{}
-	for _, c := range charts {
-		if _, found := containsVersionAndAppVersion(c.ChartVersions, version, appVersion); found {
-			result = append(result, c)
+	if cq.version != "" && cq.appVersion != "" {
+		whereClauses = append(whereClauses, fmt.Sprintf(
+			"(info->'chartVersions' @> '[{\"version\":\"%s\",\"app_version\":\"%s\"}]'::jsonb)", cq.version, cq.appVersion,
+		))
+	}
+
+	if cq.repos != nil && len(cq.repos) > 0 {
+		repoClauses := []string{}
+		for _, repo := range cq.repos {
+			if repo != "" {
+				whereQueryParams = append(whereQueryParams, repo)
+				repoClauses = append(repoClauses, fmt.Sprintf("(repo_name = $%d)", len(whereQueryParams)))
+			}
+		}
+		if len(repoClauses) > 0 {
+			repoQuery := "(" + strings.Join(repoClauses, " OR ") + ")"
+			whereClauses = append(whereClauses, repoQuery)
 		}
 	}
-	return result, nil
+	if cq.categories != nil && len(cq.categories) > 0 {
+		categoryClauses := []string{}
+		for _, category := range cq.categories {
+			if category != "" {
+				whereQueryParams = append(whereQueryParams, category)
+				categoryClauses = append(categoryClauses, fmt.Sprintf("info->>'category' = $%d", len(whereQueryParams)))
+			}
+		}
+		if len(categoryClauses) > 0 {
+			categoryQuery := "(" + strings.Join(categoryClauses, " OR ") + ")"
+			whereClauses = append(whereClauses, categoryQuery)
+		}
+	}
+	if cq.searchQuery != "" {
+		whereQueryParams = append(whereQueryParams, "%"+cq.searchQuery+"%")
+		searchClause := fmt.Sprintf("((info ->> 'name' ILIKE $%d) OR ", len(whereQueryParams)) +
+			fmt.Sprintf("(info ->> 'description' ILIKE $%d) OR ", len(whereQueryParams)) +
+			fmt.Sprintf("(info -> 'repo' ->> 'name' ILIKE $%d) OR ", len(whereQueryParams)) +
+			fmt.Sprintf("(info ->> 'keywords' ILIKE $%d) OR ", len(whereQueryParams)) +
+			fmt.Sprintf("(info ->> 'sources' ILIKE $%d) OR ", len(whereQueryParams)) +
+			fmt.Sprintf("(info -> 'maintainers' ->> 'name' ILIKE $%d))", len(whereQueryParams))
+		whereClauses = append(whereClauses, searchClause)
+	}
+	if len(whereClauses) > 0 {
+		whereQuery = "WHERE " + strings.Join(whereClauses, " AND ")
+	}
+
+	return whereQuery, whereQueryParams
 }
```

Comment:
This is the main part of this PR, the `_generateWhereClause` method. Given a `chartQuery` it generates all the WHERE clauses. 
Currently it joins every clause using AND, but we can customize it by passing the operator as a new param. If it's worthwhile, I'm willing to add it.

Reply:
I don't think it's needed to do that atm

Reply:
Also, you need to add tests for this function so you can ensure it behaves as expected. You can try with a matrix that given some inputs, produces the expected query and params.